src/libawkward/array/IndexedArray.cpp

```diff
@@ -1980,6 +1980,216 @@ namespace awkward {
     }
   }
 
+  template <typename T, bool ISOPTION>
+  const ContentPtr
+  IndexedArrayOf<T, ISOPTION>::sort_next(int64_t negaxis,
+                                         const Index64& starts,
+                                         const Index64& parents,
+                                         int64_t outlength,
+                                         bool ascending,
+                                         bool stable,
+                                         bool keepdims) const {
+    int64_t numnull;
+    struct Error err1 = util::awkward_indexedarray_numnull<T>(
+      &numnull,
+      index_.ptr().get(),
+      index_.offset(),
+      index_.length());
+    util::handle_error(err1, classname(), identities_.get());
+
+    Index64 nextparents(index_.length() - numnull);
+    Index64 nextcarry(index_.length() - numnull);
+    Index64 outindex(index_.length());
+    struct Error err2 = util::awkward_indexedarray_reduce_next_64<T>(
+      nextcarry.ptr().get(),
+      nextparents.ptr().get(),
+      outindex.ptr().get(),
+      index_.ptr().get(),
+      index_.offset(),
+      parents.ptr().get(),
+      parents.offset(),
+      index_.length());
+    util::handle_error(err2, classname(), identities_.get());
+
+    ContentPtr next = content_.get()->carry(nextcarry);
+    ContentPtr out = next.get()->sort_next(negaxis,
+                                           starts,
+                                           nextparents,
+                                           outlength,
+                                           ascending,
+                                           stable,
+                                           keepdims);
+
+    Index64 nextoutindex(index_.length());
+    int64_t j = 0;
+    for (int64_t i = 0; i < parents.length(); i++) {
+      int64_t parent = parents.getitem_at_nowrap(i);
+      int64_t start = starts.getitem_at_nowrap(parent);
+      int64_t nextparent = nextparents.getitem_at_nowrap(j);
+      if (parent == nextparent) {
+        nextoutindex.setitem_at_nowrap(i, j);
+        ++j;
+      }
+      else {
+        nextoutindex.setitem_at_nowrap(i, -1);
+      }
+    }
```

Comment:
**HERE**

This is iterating over `parents`, which is _O(n)_ in array length. What's written here is almost exactly a kernel anyway, just replacing the `getitem_at_nowrap` and `setitem_at_nowrap` with direct array access. That way, if the data are on a GPU, they won't be pulled off one at a time (I think kernel launch overhead is 10 μs); a variant of the above would run completely on the GPU, dispatched from a single function call here.

Reply:
done